runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPortablePipelineTranslator.java

```diff
@@ -0,0 +1,601 @@
+    translatorMap.put(
+        PTransformTranslation.GROUP_BY_KEY_TRANSFORM_URN,
+        FlinkBatchPortablePipelineTranslator::translateGroupByKey);
+    translatorMap.put(
+        PTransformTranslation.IMPULSE_TRANSFORM_URN,
+        FlinkBatchPortablePipelineTranslator::translateImpulse);
+    translatorMap.put(
+        PTransformTranslation.ASSIGN_WINDOWS_TRANSFORM_URN,
+        FlinkBatchPortablePipelineTranslator::translateAssignWindows);
+    translatorMap.put(
+        ExecutableStage.URN, FlinkBatchPortablePipelineTranslator::translateExecutableStage);
+    translatorMap.put(
+        PTransformTranslation.RESHUFFLE_URN,
+        FlinkBatchPortablePipelineTranslator::translateReshuffle);
+    translatorMap.put(
+        PTransformTranslation.CREATE_VIEW_TRANSFORM_URN,
+        FlinkBatchPortablePipelineTranslator::translateView);
+    return new FlinkBatchPortablePipelineTranslator(translatorMap.build());
+  }
+
+  /**
+   * Batch translation context. Stores metadata about known PCollections/DataSets and holds the
+   * flink {@link ExecutionEnvironment} that the execution plan will be applied to.
+   */
+  public static class BatchTranslationContext
+      implements FlinkPortablePipelineTranslator.TranslationContext {
+
+    private final FlinkPipelineOptions options;
+    private final ExecutionEnvironment executionEnvironment;
+    private final Map<String, DataSet<?>> dataSets;
+    private final Set<String> danglingDataSets;
+
+    private BatchTranslationContext(
+        FlinkPipelineOptions options, ExecutionEnvironment executionEnvironment) {
+      this.options = options;
+      this.executionEnvironment = executionEnvironment;
+      dataSets = new HashMap<>();
+      danglingDataSets = new HashSet<>();
+    }
+
+    @Override
+    public FlinkPipelineOptions getPipelineOptions() {
+      return options;
+    }
+
+    public ExecutionEnvironment getExecutionEnvironment() {
+      return executionEnvironment;
+    }
+
+    public <T> void addDataSet(String pCollectionId, DataSet<T> dataSet) {
+      checkArgument(!dataSets.containsKey(pCollectionId));
+      dataSets.put(pCollectionId, dataSet);
+      danglingDataSets.add(pCollectionId);
+    }
+
+    public <T> DataSet<T> getDataSetOrThrow(String pCollectionId) {
+      DataSet<T> dataSet = (DataSet<T>) dataSets.get(pCollectionId);
+      if (dataSet == null) {
+        throw new IllegalArgumentException(
+            String.format("Unknown dataset for id %s.", pCollectionId));
+      }
+
+      // Assume that the DataSet is consumed if requested. We use this as a proxy for consumption
+      // because Flink does not expose its internal execution plan.
+      danglingDataSets.remove(pCollectionId);
+      return dataSet;
+    }
+
+    public Collection<DataSet<?>> getDanglingDataSets() {
+      return danglingDataSets.stream().map(id -> dataSets.get(id)).collect(Collectors.toList());
+    }
+  }
+
+  private interface PTransformTranslator<T> {
```

Comment:
<!--new_thread; commit:c3bc9f1ff4d5f830b19aef83b758b12e363f7679; resolved:0-->
What does "T" stand for? From usage, seems that it's some kind of context? Should it be called "ContextT" then?